embassy-usb/src/lib.rs

```diff
@@ -114,6 +114,7 @@ struct Inner<'d, D: Driver<'d>> {
 
     device_state: UsbDeviceState,
     suspended: bool,
+    power_available: bool,
```

Comment:
This is a bit bike shedding, but the USB spec refers to this state as `powered` which might be a better name than `power_available`. In fact, rather than a separate `bool`, I'd add a `Unpowered` state to the `UsbDeviceState` enum. We will also need a `fn powered(&self, powered: bool)` method on the `DeviceStateHandler`.

Reply:
I have a separate bool because it represents the fact that power is set independently of `UsbDeviceState`.

Reply:
It's not possible for a USB device to be unpowered and suspended simultaneously. Once Vbus is interrupted the device needs to be reset at the very least, and usually you'll want to disable the device until Vbus is restored. Chapter 9 of the USB 2.0 spec has a nice diagram of the device state machine that might be helpful.

Reply:
Again, the signal comes from a separate source, hence my modeling it this way. Apologies if I'm misunderstanding something though. I'm not very familiar with the internals of USB.

Reply:
No worries, it's a complicated spec. I guess my main point is that we want the `embassy-usb` library to model an "ideal" USB device as closely as possible. The signals do come from separate sources, but that's an implementation detail. What we want to model is a device state that goes Unpowered -> Default -> Addressed -> Configured (<-> Suspended). In practice, that means the power removed event trumps everything else. This stuff is tricky to get right.

Here's a summary of how I think it should work:
* `Unpowered` should be another `UsbDeviceState` variant
* `run_until_suspend()` should *only* return when the `UsbDeviceState` is `Suspended`. Other state changes should be handled automatically without returning.
* `wait_for_resume()` should return as soon as the `UsbDeviceState` is anything other than `Suspended`
* A power removed event should immediately set the device state to `Unpowered` and it should remain there until a power detected event is received

The reason for the `run_until_suspend()` and `wait_for_resume()` behaviors is because the only reason those methods are split out from `run()` is because the `run()` future is not safe to drop yet we need to be able to call `remote_wakeup()` (which requires an exclusive reference to the `UsbDevice`) when the device is suspended. Therefore the behavior of those methods should be strictly tied to the `Suspended` device state.

Reply:
I've now implemented your recommendations regarding the new `Unpowered` state. Thanks so much - it looks and behaves great! Please see commit 8d71a358c8fa0524fce564357d3d36f24be41353.